keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,802 @@
+      failurePolicy: Deny
+      connectionInfo:
+        type: KubeConfig
+        kubeConfigFile: /kube-system-authz-webhook.yaml
+      matchConditions:
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests to kube-system
+      - expression: request.resourceAttributes.namespace == 'kube-system'
+      # don't intercept requests from kube-system service accounts
+      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
+  - type: Node
+  - type: RBAC
+  - type: Webhook
+    webhook:
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      failurePolicy: Deny
+      connectionInfo:
+        type: KubeConfig
+        kubeConfigFile: /opa-kube-system-authz-webhook.yaml
+      matchConditions:
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests to kube-system
+      - expression: request.resourceAttributes.namespace == 'kube-system'
+      # don't intercept requests from kube-system service accounts
+      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
+```
+
+#### Story 3: Denying requests on certain scenarios
+
+The authorizer chain should be powerful enough to deny anyone making a request
+if certain conditions are satisfied.
+
+#### Story 4: Controlling access of a privileged RBAC role
+
+Certain users associated with a privileged role might need to have their access
+scoped to certain namespaces. Having ordered authorisation modes allows the
+administrator to add a webhook restricting certain user tokens before RBAC
+grants access to the user.
+
+#### Story 5: Varying defaults across versions of the API
+
+If the default values for configuration evolve over time, affected users might
+have to override these values in case they are affected by supplying the flags
+with updated defaults. A configuration format like the one proposed can be
+versioned over time with changing defaults, mitigating this risk by only
+affecting users who knowingly bump the version.
+
+#### Story 6: Conditionally filtering requests to webhooks
+
+Let's say if an API request is being made by an user in a specific group, the
+webhook request can be skipped.
+
+
+### Risks and Mitigations
+
+- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
+each are configured. This may create inconsistencies. Mitigation is to have the
+cluster bootstrapper handle such scenarios.
+- In case an administrator enables this feature and the webhook kubeconfig file
+is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
+will not be able to start. This can be mitigated by fixing the malformed values.
+
+
+## Design Details
+
+We would like to introduce a structured file format which allows authorization
+to be configured using a flag (`--authorization-config-file`) which accepts a
```

Comment:
```suggestion
to be configured using a flag (`--authorization-config`) which accepts a
```

Reply:
This needs to be symmetric with #3332 which uses `--authentication-config`.  We are not super consistent in this space because I see both `--encryption-provider-config` and `--admission-control-config-file` as current flags to the API server.

Reply:
done